File: Easy/0021-merge-two-sorted-lists/merge2srtdlists.go
/*
 * Merges two sorted linked lists (list1 and list2) into a single sorted list.
 * Returns the head of the merged list.
 *
 * Time Complexity: O(n+m), where n is the length of list1 and m is the length of list2.
 *    - Each node is processed exactly once.
 * Space Complexity: O(1), as the merging is done in place without creating additional data structures.
 */

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val  int
 *     Next *ListNode
 * }
 */

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    // If one of the lists is empty, simply return the other
    if list1 == nil {
        return list2
    }
    if list2 == nil {
        return list1
    }

    // Create a dummy node to easily build the merged list
    dummy := &ListNode{}
    tail := dummy

    // Compare the current nodes of both lists and attach the smaller one to the merged list
    for list1 != nil && list2 != nil {
        if list1.Val <= list2.Val {
            tail.Next = list1
            list1 = list1.Next
        } else {
            tail.Next = list2
            list2 = list2.Next
        }
        tail = tail.Next
    }

    // Attach any remaining nodes from list1 or list2
    if list1 != nil {
        tail.Next = list1
    } else {
        tail.Next = list2
    }

    // Return the head of the merged list (dummy.Next)
    return dummy.Next
}

File: Easy/0021-merge-two-sorted-lists/merge2srtdlists.c
#include <stdio.h>
#include <stdlib.h>

// Function to merge two sorted lists
struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {
    if (list1 == NULL) return list2;   //Base case: if one of the lists is empty, return the other
    if (list2 == NULL) return list1; 
    struct ListNode dummy;             //Create a dummy node to simplify handling the result
    struct ListNode* tail = &dummy;    //Tail pointer to keep track of the last node in the merged list
    dummy.next = NULL;              
    while (list1 != NULL && list2 != NULL) {
        if (list1->val <= list2->val) {
            tail->next = list1;  
            list1 = list1->next; 
        } else {
            tail->next = list2;   
            list2 = list2->next;
        }
        tail = tail->next;     
    }
    //Add remaining nodes (if any) from list1 or list2
    if (list1 != NULL) {
        tail->next = list1;      
    } else {
        tail->next = list2;     
    }
    return dummy.next;             //Return the merged list, excluding the dummy node
}

// Helper function to create a new node
struct ListNode* createNode(int val) {
    struct ListNode* newNode = (struct ListNode*)malloc(sizeof(struct ListNode)); //Allocate memory for a new node
    newNode->val = val;                                                           //Set the value of the node
    newNode->next = NULL;                                                         //Initialize the next pointer to NULL
    return newNode;                                                               //Return the created node
}

/*
Complexity Analysis:

Time Complexity: O(n+m), where n is the number of nodes in list1 and m is the number of nodes in list2.
- Each node is processed exactly once, so the total number of operations is proportional to the combined length of the two lists.

Space Complexity: O(1), as the merging process is performed in place without using additional memory for a new list.
*/
